File: app/dtos/exams.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.dtos.base import BaseSerializerModel
from app.models.exams import OCRStatus


class ExamReportCreateRequest(BaseModel):
    original_filename: str
    file_path: str
    exam_date: date | None = None
    ocr_status: OCRStatus = OCRStatus.PENDING
    uploaded_at: datetime


class ExamReportResponse(BaseSerializerModel):
    id: int
    user_id: int
    original_filename: str
    file_path: str
    exam_date: date | None
    ocr_status: OCRStatus
    is_confirmed: bool
    uploaded_at: datetime
    confirmed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ExamReportUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    original_filename: str | None = None
    file_path: str | None = None
    exam_date: date | None = None
    ocr_status: OCRStatus | None = None
    uploaded_at: datetime | None = None


class ExamMeasurementCreateRequest(BaseModel):
    measurement_key: str
    measurement_name: str
    value: str | None = None
    unit: str | None = None
    ocr_confidence: Decimal | None = None
    is_user_confirmed: bool = False


class ExamMeasurementUpdateRequest(BaseModel):
    measurement_key: str | None = None
    measurement_name: str | None = None
    value: str | None = None
    unit: str | None = None
    ocr_confidence: Decimal | None = None
    is_user_confirmed: bool | None = None


class ExamMeasurementResponse(BaseSerializerModel):
    id: int
    exam_report_id: int
    measurement_key: str
    measurement_name: str
    value: str | None
    unit: str | None
    ocr_confidence: Decimal | None
    is_user_confirmed: bool
    created_at: datetime
    updated_at: datetime


class ExamReportDetailResponse(ExamReportResponse):
    measurements: list[ExamMeasurementResponse]


class ExamConfirmRequest(BaseModel):
    measurements: list[ExamMeasurementUpdateRequest] | None = None


class ExamDummyOCRResponse(BaseModel):
    message: str
    measurements: list[ExamMeasurementResponse]


class ExamOCRResponse(ExamDummyOCRResponse):
    pass

File: frontend/src/api/exams.ts
import { apiRequest } from "./client";
import type { AsyncJob } from "./jobs";

export type ExamReport = {
  id: number;
  user_id: number;
  original_filename: string;
  file_path: string;
  exam_date?: string | null;
  ocr_status: string;
  is_confirmed: boolean;
  uploaded_at: string;
  confirmed_at?: string | null;
  created_at: string;
  updated_at: string;
};

export type ExamMeasurement = {
  id: number;
  exam_report_id: number;
  measurement_key: string;
  measurement_name: string;
  value?: string | null;
  unit?: string | null;
  ocr_confidence?: string | number | null;
  is_user_confirmed: boolean;
  created_at: string;
  updated_at: string;
};

export type ExamOcrResponse = {
  message: string;
  measurements: ExamMeasurement[];
  ocr_provider?: string;
  fallback_used?: boolean;
  provider_message?: string | null;
  raw_text_preview?: string | null;
};

export async function listExams<T = ExamReport[]>(params?: { limit?: number; offset?: number }): Promise<T> {
  const query = new URLSearchParams();
  if (params?.limit != null) query.set("limit", String(params.limit));
  if (params?.offset != null) query.set("offset", String(params.offset));
  const qs = query.toString();
  return apiRequest<T>(`/exams${qs ? `?${qs}` : ""}`);
}

export async function getExam(examId: number): Promise<ExamReport> {
  return apiRequest<ExamReport>(`/exams/${examId}`);
}

export async function createExam(payload: {
  original_filename: string;
  file_path: string;
  exam_date?: string | null;
  uploaded_at: string;
}): Promise<ExamReport> {
  return apiRequest<ExamReport>("/exams", { method: "POST", body: payload });
}

export async function runExamOcr(examId: number, file: File): Promise<AsyncJob> {
  const formData = new FormData();
  formData.append("image", file);
  return apiRequest<AsyncJob>(`/exams/${examId}/ocr`, { method: "POST", body: formData });
}

export async function listMeasurements(examId: number): Promise<ExamMeasurement[]> {
  return apiRequest<ExamMeasurement[]>(`/exams/${examId}/measurements`);
}

export async function updateMeasurement(
  measurementId: number,
  payload: Partial<Pick<ExamMeasurement, "measurement_key" | "measurement_name" | "value" | "unit" | "ocr_confidence" | "is_user_confirmed">>,
): Promise<ExamMeasurement> {
  return apiRequest<ExamMeasurement>(`/exams/measurements/${measurementId}`, { method: "PATCH", body: payload });
}

export async function confirmExam(examId: number): Promise<ExamReport> {
  return apiRequest<ExamReport>(`/exams/${examId}/confirm`, { method: "POST" });
}
